File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/NotificationsApiClient.kt
/*
 * Copyright (c) 2025-present, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

package com.salesforce.androidsdk.rest

import com.salesforce.androidsdk.app.SalesforceSDKManager
import com.salesforce.androidsdk.rest.RestRequest.MEDIA_TYPE_JSON
import com.salesforce.androidsdk.rest.RestRequest.RestMethod.GET
import com.salesforce.androidsdk.rest.RestRequest.RestMethod.POST
import com.salesforce.androidsdk.util.SalesforceSDKLogger
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * Provides REST client methods for a variety of notifications API endpoints.
 * - Notifications Types: `/connect/notifications/types`
 * - Notifications Actions: `/connect/notifications/${notificationId}/actions/${actionKey}`
 *
 * See https://salesforce.quip.com/KGU3ALoXRCjK#RcfABAPLVfg
 * TODO: Replace the documentation link with the final documentation. ECJ20250310
 *
 * @param restClient The REST client to use
 */
class NotificationsApiClient(
    private val restClient: RestClient
) {

    /**
     * Submit a request to the Notifications API Types endpoint.
     * @return The endpoint's response
     */
    @Suppress("unused")
    @Throws(SfapApiException::class)
    fun fetchNotificationsTypes(): NotificationsTypesResponseBody? {
        val context = SalesforceSDKManager.getInstance().appContext
        val apiVersion = ApiVersionStrings.getVersionNumber(context)

        // Submit the request.
        if (apiVersion < "v64.0") { // TODO: Remove once MSDK default API version is 64 or greater.
            SalesforceSDKLogger.w(TAG, "Cannot request Salesforce push notifications types with API less than v64.0")
            return null
        }

        val restRequest = RestRequest(
            GET,
            "https://${restClient.clientInfo.instanceUrl.host}/${ApiVersionStrings.getBasePath()}/connect/notifications/types",
            mutableMapOf<String, String>()
        )
        val restResponse = restClient.sendSync(restRequest)
        val responseBodyString = restResponse.asString()

        return if (restResponse.isSuccess && responseBodyString != null) {
            NotificationsTypesResponseBody.fromJson(responseBodyString)
        } else {
            val errorResponseBody = responseBodyString?.let { NotificationsApiErrorResponseBody.fromJson(responseBodyString) }
            throw NotificationsApiException(
                errorCode = errorResponseBody?.firstOrNull()?.errorCode,
                message = errorResponseBody?.firstOrNull()?.message ?: "No error response body was provided by the API endpoint.",
                messageCode = errorResponseBody?.firstOrNull()?.messageCode,
                source = responseBodyString
            )
        }
    }

    /**
     * Submit a request to the Notifications API actions endpoint.
     * @return The endpoint's response
     */
    @Throws(NotificationsApiException::class)
    fun submitNotificationAction(
        notificationId: String,
        actionKey: String
    ): NotificationsActionsResponseBody? {
        val context = SalesforceSDKManager.getInstance().appContext
        val apiVersion = ApiVersionStrings.getVersionNumber(context)

        // Submit the request.
        if (apiVersion < "v64.0") { // TODO: Remove once MSDK default API version is 64 or greater.
            SalesforceSDKLogger.w(TAG, "Cannot submit Salesforce Notifications API action with API less than v64.0")
            return null
        }

        val restRequest = RestRequest(
            POST,
            "https://${restClient.clientInfo.instanceUrl.host}/${ApiVersionStrings.getBasePath()}/connect/notifications/${notificationId}/actions/${actionKey}",
            "".toRequestBody(MEDIA_TYPE_JSON),
            mutableMapOf<String, String>()
        )
        val restResponse = restClient.sendSync(restRequest)
        val responseBodyString = restResponse?.asString()

        return if (restResponse?.isSuccess == true && responseBodyString != null) {
            NotificationsActionsResponseBody.fromJson(responseBodyString)
        } else {
            val errorResponseBody = responseBodyString?.let { NotificationsApiErrorResponseBody.fromJson(responseBodyString) }
            throw NotificationsApiException(
                errorCode = errorResponseBody?.firstOrNull()?.errorCode,
                message = errorResponseBody?.firstOrNull()?.message ?: "No error response body was provided by the API endpoint.",
                messageCode = errorResponseBody?.firstOrNull()?.messageCode,
                source = responseBodyString
            )
        }
    }

    companion object {
        private const val TAG = "NotificationsApiClient"
    }
}

File: flutter_sdk/lib/src/core/rest/notifications_api_client.dart
import 'dart:convert';
import 'rest_client.dart';
import 'rest_request.dart';
import 'rest_response.dart';

/// Exception for Notifications API errors.
class NotificationsApiException implements Exception {
  final String? errorCode;
  final String? message;
  final String? messageCode;
  final String? source;

  NotificationsApiException({
    this.errorCode,
    this.message,
    this.messageCode,
    this.source,
  });

  factory NotificationsApiException.fromResponse(RestResponse response) {
    try {
      final errors = jsonDecode(response.asString);
      if (errors is List && errors.isNotEmpty) {
        final first = errors[0] as Map<String, dynamic>;
        return NotificationsApiException(
          errorCode: first['errorCode'] as String?,
          message: first['message'] as String?,
          messageCode: first['messageCode'] as String?,
          source: response.asString,
        );
      }
    } catch (_) {}
    return NotificationsApiException(
      message: 'HTTP ${response.statusCode}',
      source: response.asString,
    );
  }

  @override
  String toString() => 'NotificationsApiException: $message ($errorCode)';
}

/// Action within a notification action group.
class NotificationAction {
  final String? actionKey;
  final String? label;
  final String? name;
  final String? type;

  NotificationAction({this.actionKey, this.label, this.name, this.type});

  factory NotificationAction.fromJson(Map<String, dynamic> json) {
    return NotificationAction(
      actionKey: json['actionKey'],
      label: json['label'],
      name: json['name'],
      type: json['type'],
    );
  }

  Map<String, dynamic> toJson() => {
        'actionKey': actionKey,
        'label': label,
        'name': name,
        'type': type,
      };
}

/// Group of actions available for a notification type.
class NotificationActionGroup {
  final String? name;
  final List<NotificationAction> actions;

  NotificationActionGroup({this.name, this.actions = const []});

  factory NotificationActionGroup.fromJson(Map<String, dynamic> json) {
    return NotificationActionGroup(
      name: json['name'],
      actions: (json['actions'] as List<dynamic>?)
              ?.map((a) =>
                  NotificationAction.fromJson(a as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'name': name,
        'actions': actions.map((a) => a.toJson()).toList(),
      };
}

/// A notification type available for the current user.
class NotificationType {
  final String? apiName;
  final String? label;
  final String? type;
  final List<NotificationActionGroup> actionGroups;

  NotificationType({
    this.apiName,
    this.label,
    this.type,
    this.actionGroups = const [],
  });

  factory NotificationType.fromJson(Map<String, dynamic> json) {
    return NotificationType(
      apiName: json['apiName'],
      label: json['label'],
      type: json['type'],
      actionGroups: (json['actionGroups'] as List<dynamic>?)
              ?.map((g) =>
                  NotificationActionGroup.fromJson(g as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'apiName': apiName,
        'label': label,
        'type': type,
        'actionGroups': actionGroups.map((g) => g.toJson()).toList(),
      };
}

/// Response from the notifications types endpoint.
class NotificationsTypesResponse {
  final List<NotificationType> notificationTypes;

  NotificationsTypesResponse({this.notificationTypes = const []});

  factory NotificationsTypesResponse.fromJson(Map<String, dynamic> json) {
    return NotificationsTypesResponse(
      notificationTypes: (json['notificationTypes'] as List<dynamic>?)
              ?.map((t) =>
                  NotificationType.fromJson(t as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() => {
        'notificationTypes':
            notificationTypes.map((t) => t.toJson()).toList(),
      };
}

/// Response from the notifications actions endpoint.
class NotificationsActionsResponse {
  final String? message;

  NotificationsActionsResponse({this.message});

  factory NotificationsActionsResponse.fromJson(Map<String, dynamic> json) {
    return NotificationsActionsResponse(message: json['message']);
  }
}

/// REST client for Salesforce Notifications API endpoints (v64.0+).
///
/// Provides methods to fetch notification types and submit notification actions.
/// Mirrors Android NotificationsApiClient for feature parity.
class NotificationsApiClient {
  static const String _minApiVersion = '64.0';
  final RestClient _restClient;

  NotificationsApiClient(this._restClient);

  /// Fetches the notification types available for the current user.
  ///
  /// Returns null if API version < v64.0.
  /// Throws [NotificationsApiException] on error.
  Future<NotificationsTypesResponse?> fetchNotificationsTypes(
      String apiVersion) async {
    if (!_isVersionSupported(apiVersion)) return null;

    final request = RestRequest(
      method: RestMethod.get,
      path:
          '${RestRequest.servicesData}/v$apiVersion/connect/notifications/types',
    );

    final response = await _restClient.sendAsync(request);
    if (!response.isSuccess) {
      throw NotificationsApiException.fromResponse(response);
    }
    return NotificationsTypesResponse.fromJson(response.asJsonObject());
  }

  /// Submits an action for a notification.
  ///
  /// Returns null if API version < v64.0.
  /// Throws [NotificationsApiException] on error.
  Future<NotificationsActionsResponse?> submitNotificationAction(
    String apiVersion,
    String notificationId,
    String actionKey,
  ) async {
    if (!_isVersionSupported(apiVersion)) return null;

    final request = RestRequest(
      method: RestMethod.post,
      path:
          '${RestRequest.servicesData}/v$apiVersion/connect/notifications/$notificationId/actions/$actionKey',
    );

    final response = await _restClient.sendAsync(request);
    if (!response.isSuccess) {
      throw NotificationsApiException.fromResponse(response);
    }
    return NotificationsActionsResponse.fromJson(response.asJsonObject());
  }

  bool _isVersionSupported(String apiVersion) {
    final version = double.tryParse(apiVersion) ?? 0;
    final minVersion = double.tryParse(_minApiVersion) ?? 64.0;
    return version >= minVersion;
  }
}
